lidarts/statistics/routes.py

```diff
@@ -28,35 +28,75 @@ def x01():
+        form.computer_level.data = form.computer_level.default
+    if form.number_of_games.data is None:
+        form.number_of_games.data = custom_filter_number_of_games
+    if form.date_from.data is None:
+        form.date_from.data = custom_filter_date_from
+    if form.date_to.data is None:
+        form.date_to.data = custom_filter_date_to
 
     # check which custom filter is active
     use_custom_filter_last_games = False
     use_custom_filter_date_range = False
-    if form.selectGameRangeFilter.data == 'lastgames':
+    if form.select_game_range_filter.data == 'lastgames':
         use_custom_filter_last_games = True
-        custom_filter_number_of_games = form.numberOfGames.data
-    elif form.selectGameRangeFilter.data == 'daterange':
+        custom_filter_number_of_games = form.number_of_games.data
+    elif form.select_game_range_filter.data == 'daterange':
         use_custom_filter_date_range = True
-        custom_filter_date_from = form.dateFrom.data
-        custom_filter_date_to = form.dateTo.data
+        custom_filter_date_from = form.date_from.data
+        custom_filter_date_to = form.date_to.data
     else:
         use_custom_filter_last_games = True
 
     # select the the games (descending order - to be able to filter for last ... games)
-    games = Game.query.filter(((Game.player1 == user.id) | (Game.player2 == user.id))
-                              & (Game.status == 'completed')) \
-        .order_by(desc(Game.begin)).all()
+    query = Game.query.filter(((Game.player1 == user.id) | (Game.player2 == user.id))
+                              & (Game.status == 'completed')
+                              & (Game.type == form.game_types.data)
+                              & (Game.in_mode == form.in_mode.data)
+                              & (Game.out_mode == form.out_mode.data))
+
+    # opponent filter
+    if form.opponents.data == 'local':
+        query = query.filter(Game.opponent_type == 'local')
+    elif form.opponents.data == 'computer':
+        if form.computer_level.data == 'all':
+            # filter for all computer opponents
+            query = query.filter(Game.opponent_type.startswith('computer'))
+        else:
+            # filter for specific computer level
+            query = query.filter(Game.opponent_type == ('computer' + str(form.computer_level.data)))
+    elif form.opponents.data == 'online':
+        query = query.filter(Game.opponent_type == 'online')
+
+        # check if we must filter to games for one specific user
+        if not form.opponent_name.data is None:
+            opponent_user_name = str(form.opponent_name.data)
+            if opponent_user_name:
+                opponent_id = -1
+                # check if username is valid and if the user id is not our own user id
+                opponent_user = User.query.filter(User.username == opponent_user_name).first()
+                if (opponent_user is not None and opponent_user.id is not current_user.id):
```

Comment:
'opponent_user is not None' und nur' opponent_user' sollten äquivalent sein in dem Fall

2. Teil: is not prüft auf Objektgleichheit, was bei Integers nicht immer das macht, was du erwartest. == wäre korrekt
s. https://stackoverflow.com/questions/306313/is-operator-behaves-unexpectedly-with-integers

Reply:
pass ich an